pkg/middleware/middleware.go

```diff
@@ -40,19 +41,33 @@ func Process(ctx *quaycontext.QuayRegistryContext, quay *v1.QuayRegistry, obj cl
 		return configBundleSecret, nil
 	}
 
-	// Add TLS cert/key pair to managed `Route`
-	if routeObj, ok := obj.(*route.Route); ok &&
-		!v1.ComponentIsManaged(quay.Spec.Components, v1.ComponentTLS) &&
-		(quayComponentLabel == "quay-app-route" || quayComponentLabel == "quay-builder-route") {
-		routeObj.Spec.TLS = &route.TLSConfig{
-			Certificate: string(ctx.TLSCert),
-			Key:         string(ctx.TLSKey),
-		}
+	rt, ok := obj.(*route.Route)
+	if !ok {
+		return obj, nil
+	}
+
+	// if we are managing TLS we can simply return the original route as no change is needed.
+	if v1.ComponentIsManaged(quay.Spec.Components, v1.ComponentTLS) {
+		return obj, nil
+	}
 
-		return routeObj, nil
+	if quayComponentLabel != "quay-app-route" && quayComponentLabel != "quay-builder-route" {
+		return obj, nil
 	}
 
-	return obj, nil
+	// if we are not managing TLS then user has provided its own key pair. on this case we
+	// set up the route as passthrough thus delegating TLS termination to the pods. as quay
+	// builders route uses GRPC we do not change its route's target port.
+	rt.Spec.TLS = &route.TLSConfig{
+		Termination:                   route.TLSTerminationPassthrough,
+		InsecureEdgeTerminationPolicy: route.InsecureEdgeTerminationPolicyRedirect,
+	}
+	if quayComponentLabel == "quay-app-route" {
+		rt.Spec.Port = &route.RoutePort{
+			TargetPort: intstr.Parse("https"),
```

Comment:
I'm not sure I 100% understand this, `intstr.Parse()` is going to create an object whose `String()` method will return `"https"` and whose `IntValue()` method will return `0`?

Reply:
Yeah, it is a struct (`intstr.IntOrString`) used by k8s when a field may be an int or a string (like the port).